From 12f5f036f6d219f764ee099886d25c6afd6c9a24 Mon Sep 17 00:00:00 2001
From: Ben Gamari <ben@smart-cactus.org>
Date: Fri, 29 Oct 2021 17:06:11 -0400
Subject: [PATCH] rts/RtsSymbols: Provide a proper prototype for environ

Previously we relied on Sym_NeedsProto, but this gave the symbol a type
which conflicts with the definition that may be provided by unistd.h.

Fixes #20577.

(cherry picked from commit d90207e288c1ccafda0f5d875116efa2cf216278)
---
 rts/RtsSymbols.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/rts/RtsSymbols.c b/rts/RtsSymbols.c
index 7177399a1aa..20c8a7b2058 100644
--- a/rts/RtsSymbols.c
+++ b/rts/RtsSymbols.c
@@ -33,6 +33,11 @@
 #include <elf.h> /* _DYNAMIC */
 #endif
 
+/* We must provide a prototype for environ since depending upon the libc
+ * version it may or may not be provided by unistd.h. See #20577.
+ */
+extern char **environ;
+
 /* -----------------------------------------------------------------------------
  * Symbols to be inserted into the RTS symbol table.
  */
@@ -60,7 +65,6 @@
       SymI_HasProto(stg_sig_install)            \
       SymI_HasProto(rtsTimerSignal)             \
       SymI_HasProto_redirect(atexit, atexit, STRENGTH_STRONG) /* See Note [Strong symbols] */ \
-      SymI_NeedsDataProto(environ)              \
       SymI_NeedsDataProto(nocldstop)
 #endif
 
-- 
GitLab

